### ISSUE TITLE
Metrics support mask

### ISSUE THREAD

**Issue by YuxianMeng** [PyTorchLightning/pytorch-lightning#2580]
## 🚀 Feature

Current metrics like Accuracy/Recall would be better to support mask. 

### Motivation

For example, when I deal with a Sequence Labeling Task and pad some sequence to max-length, I do not want to calculate metrics at the padding locations.

### Pitch

I guess a simple manipulation would work for accuracy.[(here is the original one](https://github.com/PyTorchLightning/PyTorch-Lightning/blob/master/pytorch_lightning/metrics/classification.py#L24))
```
from typing import Any, Optional

import torch
from pytorch_lightning.metrics.functional.classification import (
    accuracy,
)
from pytorch_lightning.metrics.metric import TensorMetric


class MaskedAccuracy(TensorMetric):
    """
    Computes the accuracy classification score
    Example:
        >>> pred = torch.tensor([0, 1, 2, 3])
        >>> target = torch.tensor([0, 1, 2, 2])
        >>> mask = torch.tensor([1, 1, 1, 0])
        >>> metric = MaskedAccuracy(num_classes=4)
        >>> metric(pred, target, mask)
        tensor(1.)
    """

    def __init__(
        self,
        num_classes: Optional[int] = None,
        reduction: str = 'elementwise_mean',
        reduce_group: Any = None,
        reduce_op: Any = None,
    ):
        """
        Args:
            num_classes: number of classes
            reduction: a method for reducing accuracies over labels (default: takes the mean)
                Available reduction methods:
                - elementwise_mean: takes the mean
                - none: pass array
                - sum: add elements
            reduce_group: the process group to reduce metric results from DDP
            reduce_op: the operation to perform for ddp reduction
        """
        super().__init__(name='accuracy',
                         reduce_group=reduce_group,
                         reduce_op=reduce_op)
        self.num_classes = num_classes
        self.reduction = reduction

    def forward(self, pred: torch.Tensor, target: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
        """
        Actual metric computation
        Args:
            pred: predicted labels
            target: ground truth labels
            mask: only caculate metrics where mask==1
        Return:
            A Tensor with the classification score.
        """
        mask_fill = (1-mask).bool()
        pred = pred.masked_fill_(mask=mask_fill, value=-1)
        target = target.masked_fill_(mask=mask_fill, value=-1)

        return accuracy(pred=pred, target=target,
                        num_classes=self.num_classes, reduction=self.reduction)

```

**Comment by Borda** [comment]
Looks nice! @YuxianMeng want to send it as a PR?
Cc: @justusschock @SkafteNicki

**Comment by YuxianMeng** [comment]
> Looks nice! @YuxianMeng want to send it as a PR?
> Cc: @justusschock @SkafteNicki

My pleasure:) A little question is should this PR contain only masked precision metrics or also contain other metrics?

**Comment by Borda** [comment]
I would say all, in fact it would be nice to have an abstract function/class that do this masking and the new metrics would be created just its application, so for example:
- for functional make a wrapper which does the masking and all the new masked-like function will call existing functions with this wrapper
- for class make abstract class and the masked-like metrics will be created as inheriting from the mask and metric class

Does it make sense? @justusschock @SkafteNicki

**Comment by justusschock** [comment]
@YuxianMeng But with your implementation, you calculate it also for the values you set to -1 I think. 

What you instead need to do is `accuracy(pred[mask], target[mask])` which is why I wouldn't add extras for them to be honest. We can't include every special case here and masking tensors is not much overhead, which is why I'd prefer not to include this into the metrics package. Thoughts @SkafteNicki ?

**Comment by YuxianMeng** [comment]
> @YuxianMeng But with your implementation, you calculate it also for the values you set to -1 I think.
> 
> What you instead need to do is `accuracy(pred[mask], target[mask])` which is why I wouldn't add extras for them to be honest. We can't include every special case here and masking tensors is not much overhead, which is why I'd prefer not to include this into the metrics package. Thoughts @SkafteNicki ?

As for accuracy, actually only the non-negative classes are calculated. I thought about using `accuracy(pred[mask], target[mask])`, but it may [cause speed trouble when training on TPU](https://sea-region.github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md)

**Comment by SkafteNicki** [comment]
I agree with @Borda that this should be an abstract function/class. The most simple, in my opinion, would be a class that the user can wrap their already existing metric with: `masked_accuracy=MaskedMetric(Accuracy())`. This would add a additional argument to the call: `value = masked_accuracy(pred, target, mask)`. The alternative, re-writing each metric to include this feature, is not feasible at the moment.

**Comment by Borda** [comment]
@YuxianMeng mind send a PR and I guess @SkafteNicki or @justusschock could help/guide you throw 🐰

**Comment by justusschock** [comment]
I think I speak for both of us, saying that we'd for sure do that and really appreciate the PR :)

**Comment by SkafteNicki** [comment]
Yes just ping us in the PR when you are ready, and we will assist you.

**Comment by YuxianMeng** [comment]
Working on it, I will let you when I'm ready :)

**Comment by hadim** [comment]
This issue has been closed. Does the mask metrics features has landed? Or nobody has worked on it yet?

**Comment by SkafteNicki** [comment]
It was closed due to no activity, so it is still not a part of lightning. @hadim please feel free to pick it up and send a PR :]

**Comment by davzaman** [comment]
I have implemented a version of this in my own project, would anyone like to collaborate on making a PR for this?

**Comment by SkafteNicki** [comment]
@davzaman please yes, would be a great addition :]

**Comment by davzaman** [comment]
I didn't implement it as a class wrapper, but I have a few ideas on how to do it. it might take me a little while as i have deadlines for other things but i will be working on this!
